src/main/kotlin/org/rust/lang/core/types/analysis/mutability/UnusedMutability.kt

```diff
@@ -0,0 +1,186 @@
+            isImmutable
+        }
+        return isImmutable
+    }
+
+    fun check() {
+        for (path in mutabilityData.paths) {
+            if (path is UsagePath.Base && path.declaration.isMutable() && isImmutable(path)) {
+                ctx.reportNeedlesslyMutableDeclaration(path.declaration, path.declarationKind)
+            }
+        }
+    }
+
+    companion object {
+        fun buildFor(ctx: UnusedMutabilityContext, mutabilityData: UsageAnalysisData<MutabilityDataFlow>,
+                     cfg: ControlFlowGraph): FlowedMutabilityData {
+            val dfcxLivePaths = DataFlowContext(cfg, MutabilityDataFlowOperator, mutabilityData.paths.size, FlowDirection.Backward)
+
+            mutabilityData.addGenKills(dfcxLivePaths)
+            dfcxLivePaths.propagate()
+
+            return FlowedMutabilityData(ctx, mutabilityData, dfcxLivePaths)
+        }
+    }
+}
+
+private fun isMutableContext(element: RsElement): Boolean {
+    val parent = element.parent
+    val binOp = parent as? RsBinaryExpr
+    // element = ...
+    if (binOp?.isAssignBinaryExpr == true && binOp.left == element) {
+        return true
+    }
+    // &mut <element>
+    if ((parent as? RsUnaryExpr)?.operatorType == UnaryOperator.REF_MUT) {
+        return true
+    }
+    // lambda calls
+    if (parent is RsCallExpr) {
+        return true
+    }
+
+    // mutable pattern matching
+    if (parent is RsCondition) {
+        val mutable = parent.patList?.any { it.isMutable() } ?: false
+        if (mutable) return true
+    }
+    if (parent is RsMatchExpr) {
+        val mutable = parent.matchBody?.matchArmList?.any { it.patList.any { pat -> pat.isMutable() } } ?: false
+        if (mutable) return true
+    }
+
+    // let ref mut b = <element>;
+    if (parent is RsLetDecl && parent.expr == element) {
+        val binding = parent.findBinding()
+        if (binding?.kind is RsBindingModeKind.BindByReference && binding.topLevelPattern.isMutable()) {
+            return true
+        }
+    }
+
+    // &mut self method calls
+    if (element is RsDotExpr) {
+        if (element.methodCall != null) {
+            val function = element.methodCall?.reference?.resolve() as? RsFunction
+            if (function != null && function.isMethod && function.selfParameter?.mutability?.isMut == true) {
+                return true
+            }
+        }
+    }
+    if (parent is RsDotExpr && isMutableContext(parent)) {
+        return true
+    }
+
+    return false
+}
+
+private fun RsPat?.isMutable(): Boolean {
+    if (this == null) return false
+    return when (this) {
+        is RsPatRef -> mut != null
+        is RsPatTupleStruct -> patList.any { it.isMutable() }
+        is RsPatTup -> patList.any { it.isMutable() }
+        is RsPatField -> patFieldFull?.pat.isMutable() || patBinding.isMutable()
+        is RsPatIdent -> pat.isMutable() || patBinding.isMutable()
+        else -> false
+    }
+}
+
+private fun RsPatBinding?.isMutable(): Boolean {
```

Comment:
Consider using `org.rust.lang.core.psi.ext.RsPatBindingKt#getMutability` (`this.mutability.isMut`)